### DISCUSSION TITLE
Clarification on "private" modules in a package

### DISCUSSION THREAD

**Discussion by Sangeppato** [microsoft/pyright#7850] [Q&A]
In [Typing Guidance for Python Libraries](https://github.com/microsoft/pyright/blob/main/docs/typed-libraries.md) it is stated that "a type checker will treat all modules within that package as importable unless the module is marked private" and: "If a file `__init__.py` uses the form `from .A import X`, symbol “A” is not private unless the name begins with an underscore".

I have the following package structure:
```
pkg
├─ subpkg
│  ├─ __init__.py
│  └─ _a.py
├─ __init__.py
└─ b.py
```

`pkg/__init__.py`:
```
from . import b, subpkg
```

`pkg/subpkg/__init__.py`:
```
from ._a import *
```

When using autocompletion with Pylance in VSCode, I can see that imported symbols in `b.py` (such as `numpy` and `math`) are correctly considered "private" and thus not visible in the completion pop-up. On the other hand, when completing `pkg.subpkg.` I can see the submodule `_a` as importable. Maybe I didn't understand it properly, but considering the aforementioned statement I would expect `_a` to be considered private and omitted from the autocompletion suggestions. Is there anything I'm missing? Is there a way to prevent implementation modules such as `_a` from appearing in the completion pop-up? Thank you in advance.

**Comment by erictraut** [comment]
Your understanding is correct. I would expect that `_a` should not be presented as a completion suggestion when completing `from pkg.subpkg import X` (where `X` is the current location of your cursor). I recommend filling a bug in the [pylance-release](https://github.com/microsoft/pylance-release/issues) project. Since this is a language server issue, it's under the purview of the pylance team.
